base_layer/core/src/validation/chain_validators.rs

```diff
@@ -0,0 +1,150 @@
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+pub use crate::consensus::ConsensusManager;
+use crate::{
+    blocks::{
+        blockheader::{BlockHeader, BlockHeaderValidationError},
+        genesis_block::get_gen_block_hash,
+    },
+    chain_storage::{BlockchainBackend, BlockchainDatabase, MmrTree},
+    validation::{error::ValidationError, traits::Validation},
+};
+use std::sync::Arc;
+use tari_crypto::commitment::HomomorphicCommitmentFactory;
+use tari_transactions::types::CryptoFactories;
+use tari_utilities::hash::Hashable;
+
+/// This validator checks that the synced state satisfies *all* consensus rules and should only be performed on
+/// the chain tip header.
+pub struct ChainTipValidator<B: BlockchainBackend> {
+    rules: ConsensusManager<B>,
+    factories: Arc<CryptoFactories>,
+    db: BlockchainDatabase<B>,
+}
+
+impl<B: BlockchainBackend> ChainTipValidator<B>
+where B: BlockchainBackend
+{
+    pub fn new(rules: ConsensusManager<B>, factories: Arc<CryptoFactories>, db: BlockchainDatabase<B>) -> Self {
+        Self { rules, factories, db }
+    }
+
+    fn db(&self) -> Result<BlockchainDatabase<B>, ValidationError> {
+        Ok(self.db.clone())
+    }
+}
+
+impl<B: BlockchainBackend> Validation<BlockHeader, B> for ChainTipValidator<B> {
+    /// The consensus checks that are done (in order of cheapest to verify to most expensive):
+    /// 1. Does the MMR roots calculated from the chain state match that roots provided in the tip header?
+    /// 1. Is the full accounting balance of the current chain state valid?
+    fn validate(&self, block_header: &BlockHeader) -> Result<(), ValidationError> {
+        check_mmr_roots(block_header, self.db()?)?;
+        check_chain_accounting_balance(block_header, self.db()?, self.rules.clone(), &self.factories)?;
+
+        Ok(())
+    }
+}
+
+/// This validator checks that the synced chain builds on the correct genesis block and should only be performed on the
+/// genesis block header.
+pub struct ChainGbValidator {}
```

Comment:
```suggestion
pub struct GenesisBlockValidator {}
```

I had no idea what this was supposed to be until I got here

Reply:
Change applied.